actions.py

```diff
@@ -0,0 +1,62 @@
+import pickle
+import base64
+from pathlib import Path
+import subprocess
+
+from flask import Blueprint, request, jsonify, session
+
+bp = Blueprint("actions", __name__)
+
+
+@bp.route("/message", methods=["POST"])
+def log_entry():
+    user_info = session.get("user_info", None)
+    if user_info is None:
+        return jsonify({"error": "no user_info found in session"})
+    access_level = user_info[2]
+    if access_level > 2:
+        return jsonify({"error": "access level < 2 is required for this action"})
+    filename_param = request.form.get("filename")
+    if filename_param is None:
+        return jsonify({"error": "filename parameter is required"})
+    text_param = request.form.get("text")
+    if text_param is None:
+        return jsonify({"error": "text parameter is required"})
+
+    user_id = user_info[0]
+    user_dir = "data/" + str(user_id)
+    user_dir_path = Path(user_dir)
+    if not user_dir_path.exists():
+        user_dir_path.mkdir()
+
+    filename = filename_param + ".txt"
+    path = Path(user_dir + "/" + filename)
+    # vulnerability: Directory Traversal
+    with path.open("w", encoding="utf-8") as open_file:
+        open_file.write(text_param)
+    return jsonify({"success": True})
+
+
+@bp.route("/grep_processes")
+def grep_processes():
+    name = request.args.get("name")
+    # vulnerability: Remote Code Execution
+    res = subprocess.run(
+        ["ps aux | grep " + name + " | awk '{print $11}'"],
+        shell=True,
+        capture_output=True,
+    )
```

Comment:
<div align="center"> 

### 🎯 Static Application Security Testing (SAST) Vulnerability 
	
| Severity | Finding |
| :---: | :---: |
| ![](https://raw.githubusercontent.com/jfrog/frogbot/master/resources/v2/applicableLowSeverity.png)<br> | Parameter Injection |

#### 👇 Details

<details>
<summary> <b>Full description</b> </summary>
<br>


In a Parameter Injection vulnerability, the attacker may control a parameter passed
to a child process. The severity of the issue is lower than command injection, and
its impact will depend on the specifics of the child process being run.
### Vulnerable example
```python
    @app.route("/start_proc", methods=["POST"])
    def calculate():
        param_option = request.form.get("param")
        subprocess.run(["/opt/proc/proc", param_option])
```
### Remediation
Avoid passing non-whitelisted arguments to child processes
```diff
-   subprocess.run(["/opt/proc/proc", param_option])
+   if param_option in whitelist:
+       subprocess.run(["/opt/proc/proc", whitelist[param_option]])
```


</details>



<details>
<summary><b>Code Flows</b> </summary>



<details>
<summary><b>1. Vulnerable data flow analysis result</b> </summary>
<br>

↘️. `request.args` (at actions.py line 42)

↘️. `request.args.get` (at actions.py line 42)

↘️. `request.args.get("name")` (at actions.py line 42)

↘️. `name` (at actions.py line 42)

↘️. `"ps aux | grep " + name` (at actions.py line 45)

↘️. `"ps aux | grep " + name + " | awk '{print $11}'"` (at actions.py line 45)

↘️. `["ps aux | grep " + name + " | awk '{print $11}'"]` (at actions.py line 45)

↘️. `subprocess.run(
        ["ps aux | grep " + name + " | awk '{print $11}'"],
        shell=True,
        capture_output=True,
    )` (at actions.py line 44)


</details>



</details>


---
<div align="center">

[🐸 JFrog Frogbot](https://github.com/jfrog/frogbot#readme)

</div>